iroh-relay/src/quic.rs

```diff
@@ -0,0 +1,301 @@
+//! Create a QUIC server that accepts connections
+//! for QUIC address discovery.
+use std::{net::SocketAddr, sync::Arc};
+
+use anyhow::Result;
+use quinn::VarInt;
+
+/// ALPN for our quic addr discovery
+pub const ALPN_QUIC_ADDR_DISC: &[u8] = b"quic";
+/// Endpoint close error code
+pub const QUIC_ADDR_DISC_CLOSE_CODE: VarInt = VarInt::from_u32(0);
+/// Endpoint close reason
+pub const QUIC_ADDR_DISC_CLOSE_REASON: &[u8] = b"finished";
+
+#[cfg(feature = "server")]
+pub(crate) mod server {
+    use quinn::ApplicationClose;
+    use tokio::task::JoinSet;
+    use tokio_util::{sync::CancellationToken, task::AbortOnDropHandle};
+    use tracing::{debug, info, info_span, warn, Instrument};
+
+    use super::*;
+    pub use crate::server::QuicConfig;
+
+    pub struct QuicServer {
+        bind_addr: SocketAddr,
+        cancel: CancellationToken,
+        handle: AbortOnDropHandle<()>,
+    }
+
+    impl QuicServer {
+        /// Returns a handle for this server.
+        ///
+        /// The server runs in the background as several async tasks.  This allows controlling
+        /// the server, in particular it allows gracefully shutting down the server.
+        pub fn handle(&self) -> ServerHandle {
+            ServerHandle {
+                cancel_token: self.cancel.clone(),
+            }
+        }
+
+        /// Returns the [`AbortOnDropHandle`] for the supervisor task managing the endpoint.
+        ///
+        /// This is the root of all the tasks for the QUIC address discovery service.  Aborting it will abort all the
+        /// other tasks for the service.  Awaiting it will complete when all the service tasks are
+        /// completed.[]
+        pub fn task_handle(&mut self) -> &mut AbortOnDropHandle<()> {
+            &mut self.handle
+        }
+
+        /// Returns the socket address for this QUIC server.
+        pub fn bind_addr(&self) -> &SocketAddr {
+            &self.bind_addr
+        }
+
+        /// Spawns a QUIC server that creates and QUIC endpoint and listens
+        /// for QUIC connections for address discovery
+        ///
+        /// # Panics
+        /// If there is a panic during a connection, it will be propagated
+        /// up here. Any other errors in a connection will be logged as a
+        ///  warning.
+        pub(crate) fn spawn(quic_config: QuicConfig) -> Result<Self> {
+            let mut server_config =
+                quinn::ServerConfig::with_crypto(Arc::new(quic_config.server_config));
+            let transport_config = Arc::get_mut(&mut server_config.transport).unwrap();
+            transport_config
+                .max_concurrent_uni_streams(0_u8.into())
```

Comment:
Should we also disable `max_concurrent_bidi_streams`?  That would avoid another set of buffers that can be filled by a malicious remote.  And my understanding from the protocol is that no data is ever sent.